great_expectations/experimental/datasources/multi_model.py

```diff
@@ -0,0 +1,505 @@
+
+        LOGGER.debug(f"Loaded 'assets' ->\n{repr(loaded_assets)}")
+        return loaded_assets
+
+    def _execution_engine_type(self) -> Type[ExecutionEngine]:
+        """Returns the execution engine to be used"""
+        return self.execution_engine_override or self.execution_engine_type()
+
+    def execution_engine_type(self) -> Type[ExecutionEngine]:
+        """Return the ExecutionEngine type use for this Datasource"""
+        raise NotImplementedError(
+            "One needs to implement 'execution_engine_type' on a Datasource subclass"
+        )
+
+
+@runtime_checkable
+class Datasource(Protocol):
+    # instance attrs
+    name: str
+    assets: Dict[str, DataAsset]
+    execution_engine: ExecutionEngine
+    config: DatasourceConfig
+
+    @overload
+    def __init__(
+        self,
+        config: DatasourceConfig,
+    ) -> None:
+        ...
+
+    @overload
+    def __init__(
+        self,
+        config: None = ...,
+        type: str = ...,
+        name: str = ...,
+        execution_engine: ExecutionEngine = ...,
+        assets: Optional[Dict[str, DataAssetConfig]] = ...,
+        **kwargs,
+    ) -> None:
+        ...
+
+    def __init__(
+        self,
+        config: Optional[DatasourceConfig] = None,
+        type: Optional[str] = None,
+        name: Optional[str] = None,
+        execution_engine: Optional[ExecutionEngine] = None,
+        assets: Optional[Dict[str, DataAssetConfig]] = None,
+        **kwargs,
+    ) -> None:
+        """
+        Config validation errors are surfaced when trying to initialize the `DatasourceConfig`.
+
+        Optionally we could also leverage the type coercion of `pydantic.validate_arguments` to coerce compatible `dict`
+        objects into `DatasourceConfig` objects. Relying on `pydantic` to throw validation errors as appropriate.
+
+        This logic can be simplified if we only accept either a `DatasourceConfig` or keyword arguments but not both.
+        """
+        if isinstance(config, DatasourceConfig):
+            self.config = config
+        else:
+            self.config = DatasourceConfig(
+                type=type,
+                name=name,
+                assets=assets,
+                execution_engine=execution_engine,
+                **kwargs,
+            )
+
+        self.name = self.config.name
+        self.assets: Dict[str, DataAsset] = {}
+
+        # convert the DataAsset configs into runtime objects
+        for asset_name, asset_config in self.config.assets.items():
+            self.assets[asset_name] = self.asset_factory(asset_config, self)  # type: ignore[arg-type] # mypy confused about singledispatch instance method?
+
+    @functools.singledispatch
+    def asset_factory(self, asset_config: DataAssetConfig) -> DataAsset:
+        """Creates runtime asset objects from asset config models."""
+        raise NotImplementedError("No single dispatch functions have been registered")
```

Comment:
### asset_factory() single-dispatch

Now we have to create the correct runtime asset type according to the asset config models.
Rather than having to maintain another mapping or lookup table. Each asset type registers a `@single_dispatch` factory method that accepts a particular `ConfigModel` and returns a `RuntimeModel`.